File: ports-js/parser.js
import { readFileSync } from 'fs';
import URL from 'url';


function tokenize(expression) {
    const pattern = /\s*(,@|[('`,)]|"(?:[\\].|[^\\"])*"|;.*|[^\s('"`;,)]*)(.*)/;
    let matcher;
    const tokens = [];
    let part;
    
    for (const line of expression.split('\n')) {
        part = line;
        matcher = part.match(pattern);
        while (matcher && matcher[0] !== '') {
            const token = matcher[1];
            if (token && token !== '' && !token.startsWith(';')) {
                tokens.push(token);
            }
            part = matcher[2];
            matcher = part.match(pattern);
        }
    }
    return tokens;
};

function parseTokens(tokens) {
    if (tokens.length === 0) {
        return [];
    }

    const token = tokens.shift();
    if (token === '(') {
        const list = [];
        while (tokens[0] !== ')') {
            list.push(parseTokens(tokens));
        }
        tokens.shift();
        return list;
    } else if (token === ')') {
        throw new Error("Unexpected ')'");
    } else if (token === "'") {
        return [Symbol.for("quote"), parseTokens(tokens)];
    } else if (token === "`") {
        return [Symbol.for("quasiquote"), parseTokens(tokens)];
    } else if (token === ",") {
        return [Symbol.for("unquote"), parseTokens(tokens)];
    } else if (token === ",@") {
        return [Symbol.for("unquote-splicing"), parseTokens(tokens)];
    } else {
        return parseAtom(token);
    }
};

function parseAtom(token) {
    // 'Numbers become numbers; #t and #f are booleans; "..." string; otherwise Symbol.'
    const lowerToken = token.toLowerCase();
    if (lowerToken === '#t' || lowerToken === '#true') {
        return true;
    }
    if (lowerToken === '#f' || lowerToken === '#false') {
        return false;
    }
    if (token[0] === '"') {
        const rawString = token.slice(1, -1);
        return rawString.replace(/\\n/g, '\n')
            .replace(/\\r/g, '\r')
            .replace(/\\t/g, '\t');
    }
    
    let numberParseTry = parseInt(token);
    if (!isNaN(numberParseTry) && isFinite(token)) {
        return numberParseTry;
    }
    numberParseTry = parseFloat(token); // TODO: This should try to parse a fraction instead
    if (!isNaN(numberParseTry) && isFinite(token)) {
        return numberParseTry; // Replace Fraction with native float for now
    } 
    return Symbol.for(token); // Use native JavaScript Symbol.for()
};

export function parseWithoutExpand(inputString) {
    const tokens = tokenize(inputString);
    return parseTokens(tokens);
};

function main() {
    const testTable = JSON.parse(readFileSync('ports/syntax-tests.json'));

    function matches(structure, target) {
        if (Array.isArray(target)) {
            if (!Array.isArray(structure)) {
                return false;
            }
            if (structure.length !== target.length) {
                return false;
            } else {
                let result = true;
                for (let i = 0; i < target.length; i++) {
                    result = result && matches(structure[i], target[i]);
                }
                return result;
            }
        } else if (target === "Boolean") {
            return typeof structure === 'boolean';
        } else if (target === "String") {
            return typeof structure === 'string';
        } else if (target === "Character") {
            return typeof structure === 'string' && structure.length === 1;
        } else if (target === "Symbol") {
            return typeof structure === 'symbol';
        } else if (target === "Number") {
            return typeof structure === 'number';
        }
    }
    
    for (const entry of testTable.filter(row => typeof row !== 'string')) {
        const parseResult = parseWithoutExpand(entry[0]);
        if (matches(parseResult, entry[1])) {
            console.log(`✅: ${JSON.stringify(entry)}`);
        } else {
            console.log(`❌: ${JSON.stringify(entry)} got ${JSON.stringify(parseResult)} instead`);
        }
    }     
    console.log("End of test run");
}

// run tests only if started directly
if (import.meta.url.startsWith('file:')) {
    const modulePath = URL.fileURLToPath(import.meta.url);
    if (process.argv[1] === modulePath) {
        main()
    }
  }

File: ports-r/parser.rb
require 'json'

def tokenize(expression)
  pattern = /\s*(,@|[('`,)]|"(?:\\.|[^\\"])*"|;.*|[^\s('"`;,)]*)(.*)/
  tokens = []
  
  expression.split("\n").each do |line|
    part = line
    while match = part.match(pattern) and match[0] != ''
      token = match[1]
      if token && token != '' && !token.start_with?(';')
        tokens << token
      end
      part = match[2]
    end
  end
  return tokens
end

def parse_tokens(tokens)
  return [] if tokens.empty?

  token = tokens.shift
  if token == '('
    list = []
    while tokens[0] != ')'
      list << parse_tokens(tokens)
    end
    tokens.shift # Remove ')'
    return list
  elsif token == ')'
    raise "Unexpected ')'"
  elsif token == "'"
    return [:quote, parse_tokens(tokens)]
  elsif token == "`"
    return [:quasiquote, parse_tokens(tokens)]
  elsif token == ","
    return [:unquote, parse_tokens(tokens)]
  elsif token == ",@"
    return [:"unquote-splicing", parse_tokens(tokens)]
  else
    return parse_atom(token)
  end
end

def parse_atom(token)
  lower_token = token.downcase
  if lower_token == '#t' || lower_token == '#true'
    return true
  end
  if lower_token == '#f' || lower_token == '#false'
    return false
  end
  if token[0] == '"'
    raw_string = token[1..-2]
    return raw_string.gsub(/\\n/, "\n")
                     .gsub(/\\r/, "\r")
                     .gsub(/\\t/, "\t")
  end
  
  # Try to parse as integer
  begin
    integer = Integer(token)
    return integer
  rescue ArgumentError
    # Not an integer
  end
  
  # Try to parse as float
  begin
    float = Float(token)
    return float
  rescue ArgumentError
    # Not a number
  end
  
  return token.to_sym  # Use Ruby's built-in symbols
end

def parse_without_expand(input_string)
  tokens = tokenize(input_string)
  return parse_tokens(tokens)
end

def matches(structure, target)
  if target.is_a?(Array)
    return false unless structure.is_a?(Array)
    return false unless structure.length == target.length
    
    result = true
    target.each_with_index do |t, i|
      result = result && matches(structure[i], t)
    end
    return result
  elsif target == "Boolean"
    return structure == true || structure == false
  elsif target == "String"
    return structure.is_a?(String)
  elsif target == "Character"
    return structure.is_a?(String) && structure.length == 1
  elsif target == "Symbol"
    return structure.is_a?(Symbol)  # Check against Ruby's built-in Symbol class
  elsif target == "Number"
    return structure.is_a?(Numeric)
  end
end

if __FILE__ == $PROGRAM_NAME
  test_table = JSON.parse(File.read('ports/syntax-tests.json'))
  
  test_table.each do |entry|
    if entry.is_a?(String)
      puts "\n"
      puts entry
      next
    end
    
    parse_result = parse_without_expand(entry[0])
    if matches(parse_result, entry[1])
      puts "✅: #{entry}"
    else
      puts "❌: #{entry} got #{parse_result.inspect} instead"
    end
  end
  puts "End of test run"
end
